File: ios/Classes/HTKLinePriceLine.swift
//
//  HTKLinePriceLine.swift
//  HTKLineView
//
//  Created by Codex on 2026/01/28.
//

import UIKit

struct HTKLinePriceLine {
    let id: String
    let value: CGFloat
    let color: UIColor
    let width: CGFloat
    let dash: [CGFloat]
    let label: String
    let labelPosition: String
    let labelColor: UIColor
    let visible: Bool

    static func packModelArray(
        _ modelList: [[String: Any]],
        defaultColor: UIColor,
        defaultLabelColor: UIColor,
        defaultWidth: CGFloat
    ) -> [HTKLinePriceLine] {
        var lines = [HTKLinePriceLine]()
        for dictionary in modelList {
            let id = dictionary["id"] as? String ?? ""
            let rawValue = dictionary["price"] ?? dictionary["value"]
            guard let number = rawValue as? NSNumber else {
                continue
            }
            let value = CGFloat(truncating: number)
            let colorValue = dictionary["color"] as? Int
            let color = (colorValue != nil ? RCTConvert.uiColor(colorValue) : nil) ?? defaultColor
            let widthValue = dictionary["width"] as? NSNumber
            let width = widthValue != nil ? CGFloat(truncating: widthValue!) : defaultWidth
            let dashList = dictionary["dash"] as? [NSNumber] ?? []
            let dash = dashList.map { CGFloat(truncating: $0) }
            let label = dictionary["label"] as? String ?? ""
            let labelPosition = dictionary["labelPosition"] as? String ?? "right"
            let labelColorValue = dictionary["labelColor"] as? Int
            let labelColor = (labelColorValue != nil ? RCTConvert.uiColor(labelColorValue) : nil) ?? defaultLabelColor
            let visible = dictionary["visible"] as? Bool ?? true

            lines.append(
                HTKLinePriceLine(
                    id: id,
                    value: value,
                    color: color,
                    width: width,
                    dash: dash,
                    label: label,
                    labelPosition: labelPosition,
                    labelColor: labelColor,
                    visible: visible
                )
            )
        }
        return lines
    }
}

File: android/src/main/java/com/github/fujianlian/klinechart/HTKLinePriceLine.java
package com.github.fujianlian.klinechart;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class HTKLinePriceLine {
    public String id;
    public float value;
    public int color;
    public float width;
    public float[] dash;
    public String label;
    public String labelPosition;
    public int labelColor;
    public boolean visible = true;

    public static List<HTKLinePriceLine> packModelArray(
            List modelList,
            int defaultColor,
            int defaultLabelColor,
            float defaultWidth
    ) {
        List<HTKLinePriceLine> lines = new ArrayList<>();
        if (modelList == null) {
            return lines;
        }
        for (Object object : modelList) {
            if (!(object instanceof Map)) {
                continue;
            }
            Map<String, Object> map = (Map<String, Object>) object;
            Object rawValue = map.get("price");
            if (!(rawValue instanceof Number)) {
                rawValue = map.get("value");
            }
            if (!(rawValue instanceof Number)) {
                continue;
            }
            HTKLinePriceLine entry = new HTKLinePriceLine();
            Object idObj = map.get("id");
            entry.id = idObj != null ? idObj.toString() : "";
            entry.value = ((Number) rawValue).floatValue();
            Object colorObj = map.get("color");
            entry.color = colorObj instanceof Number ? ((Number) colorObj).intValue() : defaultColor;
            Object widthObj = map.get("width");
            entry.width = widthObj instanceof Number ? ((Number) widthObj).floatValue() : defaultWidth;
            Object dashObj = map.get("dash");
            if (dashObj instanceof List) {
                List dashList = (List) dashObj;
                float[] dash = new float[dashList.size()];
                for (int i = 0; i < dashList.size(); i++) {
                    Object dashValue = dashList.get(i);
                    dash[i] = dashValue instanceof Number ? ((Number) dashValue).floatValue() : 0f;
                }
                entry.dash = dash;
            } else {
                entry.dash = new float[0];
            }
            Object labelObj = map.get("label");
            entry.label = labelObj != null ? labelObj.toString() : "";
            Object labelPositionObj = map.get("labelPosition");
            entry.labelPosition = labelPositionObj != null ? labelPositionObj.toString() : "right";
            Object labelColorObj = map.get("labelColor");
            entry.labelColor = labelColorObj instanceof Number ? ((Number) labelColorObj).intValue() : defaultLabelColor;
            Object visibleObj = map.get("visible");
            entry.visible = !(visibleObj instanceof Boolean) || (Boolean) visibleObj;

            lines.add(entry);
        }
        return lines;
    }
}
